File: android_app/src/main/java/me/timeto/app/ui/ActivityColorPickerSheet.kt
package me.timeto.app.ui

import androidx.activity.compose.BackHandler
import androidx.compose.animation.*
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Done
import androidx.compose.material.icons.rounded.ExpandCircleDown
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.goldenRatioDown
import me.timeto.app.onePx
import me.timeto.app.rememberVM
import me.timeto.app.toColor
import me.timeto.shared.ColorRgba
import me.timeto.shared.vm.ActivityColorPickerSheetVM

private val circleSize = 40.dp
private val circlePadding = 4.dp
private val sheetHPadding = MyListView.PADDING_OUTER_HORIZONTAL
private val dividerPadding = sheetHPadding.goldenRatioDown()

@Composable
fun ActivityColorPickerSheet(
    layer: WrapperView.Layer,
    initData: ActivityColorPickerSheetVM.InitData,
    onPick: (ColorRgba) -> Unit,
) {

    val (vm, state) = rememberVM { ActivityColorPickerSheetVM(initData) }

    Column(
        modifier = Modifier
            .fillMaxHeight()
            .background(c.background2)
            .navigationBarsPadding()
    ) {

        val circleScrollState = rememberScrollState()
        val activitiesScrollState = rememberScrollState()

        Sheet.HeaderView(
            onCancel = { layer.close() },
            title = state.headerTitle,
            doneText = state.doneTitle,
            isDoneEnabled = true,
            scrollToHeader = (circleScrollState.value + activitiesScrollState.value) * 4,
            bgColor = c.background2,
            dividerColor = c.dividerBg2,
            maxLines = 1,
        ) {
            onPick(state.selectedColor)
            layer.close()
        }

        Row(
            modifier = Modifier
                .weight(1f),
        ) {

            Row(
                modifier = Modifier
                    .verticalScroll(state = activitiesScrollState)
                    .padding(top = circlePadding)
                    .padding(start = sheetHPadding)
                    .height(IntrinsicSize.Max)
                    .weight(1f),
            ) {

                Column(
                    modifier = Modifier
                        .padding(end = dividerPadding)
                        .weight(1f)
                ) {

                    Text(
                        text = state.title,
                        modifier = Modifier
                            .padding(top = 1.dp)
                            .height(circleSize - 2.dp)
                            .clip(MySquircleShape(len = 70f))
                            .background(state.selectedColor.toColor())
                            .wrapContentHeight(Alignment.CenterVertically)
                            .padding(start = 11.dp, end = 13.dp),
                        color = c.white,
                        fontWeight = FontWeight.Medium,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )

                    Text(
                        text = state.otherActivitiesTitle,
                        modifier = Modifier
                            .padding(start = 4.dp, top = 24.dp),
                        fontSize = 11.sp,
                        fontWeight = FontWeight.Medium,
                        color = c.gray1,
                    )

                    state.allActivities.forEach { activityUI ->
                        Text(
                            text = activityUI.text,
                            modifier = Modifier
                                .padding(top = 8.dp)
                                .clip(MySquircleShape())
                                .background(activityUI.colorRgba.toColor())
                                .padding(start = 7.dp, end = 8.dp, top = 4.dp, bottom = 4.5.dp),
                            fontSize = 13.sp,
                            color = c.white,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                        )
                    }

                    Box(Modifier.navigationBarsPadding())
                }

                Box(
                    Modifier
                        .width(onePx)
                        .padding(top = circlePadding)
                        .navigationBarsPadding()
                        .fillMaxHeight()
                        .background(c.dividerBg2)
                )
            }

            Column(
                modifier = Modifier
                    .verticalScroll(state = circleScrollState)
                    .padding(
                        start = dividerPadding - circlePadding,
                        end = sheetHPadding - circlePadding,
                    ),
            ) {

                state.colorGroups.forEach { colors ->
                    Row {
                        colors.forEach { colorItem ->
                            Row(
                                modifier = Modifier
                                    .size(circleSize + (circlePadding * 2))
                                    .padding(circlePadding)
                                    .clip(RoundedCornerShape(99.dp))
                                    .background(colorItem.colorRgba.toColor())
                                    .clickable {
                                        vm.upColorRgba(colorItem.colorRgba)
                                    },
                                horizontalArrangement = Arrangement.Center,
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                AnimatedVisibility(
                                    visible = colorItem.isSelected,
                                    enter = fadeIn(),
                                    exit = fadeOut(),
                                ) {
                                    Icon(
                                        Icons.Rounded.Done,
                                        contentDescription = "Selected",
                                        modifier = Modifier
                                            .size(24.dp),
                                        tint = c.white,
                                    )
                                }
                            }
                        }
                    }
                }

                Text(
                    text = "Custom",
                    modifier = Modifier
                        .padding(bottom = 20.dp)
                        .navigationBarsPadding()
                        .clip(MySquircleShape())
                        .clickable {
                            vm.toggleIsRgbSlidersShowed()
                        }
                        .padding(horizontal = circlePadding, vertical = 2.dp),
                    color = c.blue,
                    fontSize = 14.sp,
                )
            }
        }

        DividerBg2()

        Column {

            BackHandler(state.isRgbSlidersShowed) {
                vm.toggleIsRgbSlidersShowed()
            }

            AnimatedVisibility(
                visible = state.isRgbSlidersShowed,
                enter = expandVertically(spring(stiffness = Spring.StiffnessMedium))
                        + fadeIn(spring(stiffness = Spring.StiffnessMedium)),
                exit = shrinkVertically(spring(stiffness = Spring.StiffnessMedium))
                       + fadeOut(spring(stiffness = Spring.StiffnessHigh)),
            ) {
                Column(
                    modifier = Modifier
                        .background(c.background2)
                        .pointerInput(Unit) { },
                ) {

                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(top = 10.dp, bottom = 4.dp),
                    ) {

                        Text(
                            text = state.rgbText,
                            modifier = Modifier
                                .align(Alignment.Center)
                                .clip(MySquircleShape())
                                .background(state.selectedColor.toColor())
                                .padding(vertical = 4.dp, horizontal = 8.dp),
                            color = c.white,
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Light,
                        )

                        Icon(
                            Icons.Rounded.ExpandCircleDown,
                            "Hide",
                            tint = c.iconButtonBg1,
                            modifier = Modifier
                                .align(Alignment.CenterEnd)
                                .padding(end = sheetHPadding)
                                .size(30.dp)
                                .clip(RoundedCornerShape(99.dp))
                                .background(c.background2)
                                .clickable {
                                    vm.toggleIsRgbSlidersShowed()
                                }
                        )
                    }

                    ColorSliderView(state.r, c.red, state.isRgbSlidersAnimated) { vm.upR(it) }
                    ColorSliderView(state.g, c.green, state.isRgbSlidersAnimated) { vm.upG(it) }
                    ColorSliderView(state.b, c.blue, state.isRgbSlidersAnimated) { vm.upB(it) }
                }
            }
        }
    }
}

@Composable
private fun ColorSliderView(
    value: Float,
    color: Color,
    isAnimated: Boolean,
    onChange: (Float) -> Unit,
) {
    val animatedValue = animateFloatAsState(value)
    Slider(
        // Animation works bad with manual slide
        value = if (isAnimated) animatedValue.value else value,
        onValueChange = { onChange(it) },
        modifier = Modifier.padding(horizontal = sheetHPadding - 4.dp),
        valueRange = 0f..255f,
        colors = SliderDefaults.colors(
            thumbColor = color,
            activeTrackColor = color,
        )
    )
}

File: ios_app/iphone/ActivityColorPickerSheet.swift
import SwiftUI
import shared

private let circleSize = 42.0
private let circlePadding = 4.0
private let circleCellSize = circleSize + (circlePadding * 2.0)
private let sheetHPadding = MyListView.PADDING_OUTER_HORIZONTAL
private let dividerPadding = sheetHPadding.goldenRatioDown()

struct ActivityColorPickerSheet: View {

    @State private var vm: ActivityColorPickerSheetVM
    @Binding private var isPresented: Bool
    private let onPick: (ColorRgba) -> Void

    @State private var circlesScroll = 0
    @State private var activitiesScroll = 0

    @State private var isRgbSlidersShowed = false

    init(
            isPresented: Binding<Bool>,
            initData: ActivityColorPickerSheetVM.InitData,
            onPick: @escaping (ColorRgba) -> Void
    ) {
        vm = ActivityColorPickerSheetVM(initData: initData)
        _isPresented = isPresented
        self.onPick = onPick
    }

    var body: some View {

        VMView(vm: vm, stack: .VStack()) { state in

            SheetHeaderView(
                    onCancel: { isPresented.toggle() },
                    title: state.headerTitle,
                    doneText: state.doneTitle,
                    isDoneEnabled: true,
                    scrollToHeader: circlesScroll + activitiesScroll
            ) {
                onPick(state.selectedColor)
                isPresented = false
            }

            VStack {

                HStack {

                    ScrollViewWithVListener(showsIndicators: false, vScroll: $activitiesScroll) {

                        HStack {

                            VStack(alignment: .leading) {

                                Text(state.title)
                                        .font(.system(size: 17, weight: .bold))
                                        .foregroundColor(.white)
                                        .lineLimit(1)
                                        .padding(.leading, 11)
                                        .padding(.trailing, 13)
                                        .frame(height: circleSize - 4)
                                        .background(
                                                RoundedRectangle(cornerRadius: 12, style: .continuous)
                                                        .fill(state.selectedColor.toColor())
                                        )
                                        .padding(.top, 2)

                                Text(state.otherActivitiesTitle)
                                        .foregroundColor(.secondary)
                                        .fontWeight(.medium)
                                        .font(.system(size: 13))
                                        .padding(.leading, 4)
                                        .padding(.top, 28)

                                ForEach(state.allActivities, id: \.self) { activityUI in
                                    Text(activityUI.text)
                                            .font(.system(size: 15, weight: .medium))
                                            .foregroundColor(.white)
                                            .lineLimit(1)
                                            .padding(.leading, 9)
                                            .padding(.trailing, 10)
                                            .padding(.top, 6)
                                            .padding(.bottom, 6)
                                            .background(
                                                    RoundedRectangle(cornerRadius: 10, style: .continuous)
                                                            .fill(activityUI.colorRgba.toColor())
                                            )
                                            .padding(.top, 8)
                                }

                                Padding(vertical: 20)
                            }
                                    .frame(maxWidth: .infinity, alignment: .leading)
                                    .padding(.leading, sheetHPadding)
                                    .padding(.trailing, dividerPadding)

                            ZStack {}
                                    .frame(width: onePx)
                                    .frame(maxHeight: .infinity)
                                    .background(Color(.systemGray4))
                                    .padding(.top, circlePadding)
                        }
                                .padding(.top, circlePadding)
                                .safeAreaPadding(.bottom)
                    }

                    ScrollViewWithVListener(showsIndicators: false, vScroll: $circlesScroll) {

                        VStack(alignment: .leading) {

                            ForEach(state.colorGroups, id: \.self) { colors in

                                HStack {

                                    ForEach(colors, id: \.self) { colorItem in

                                        Button(
                                                action: {
                                                    vm.upColorRgba(colorRgba: colorItem.colorRgba)
                                                },
                                                label: {

                                                    ZStack {

                                                        Circle()
                                                                .foregroundColor(colorItem.colorRgba.toColor())
                                                                .frame(width: circleSize, height: circleSize)

                                                        if colorItem.isSelected {
                                                            Image(systemName: "checkmark")
                                                                    .font(.system(size: 18, weight: .medium))
                                                                    .foregroundColor(.white)
                                                        }
                                                    }
                                                            .padding(.all, circlePadding)
                                                }
                                        )
                                                .frame(width: circleCellSize, height: circleCellSize)
                                    }
                                }
                            }

                            Button(
                                    action: {
                                        vm.toggleIsRgbSlidersShowed()
                                    },
                                    label: { Text("Custom") }
                            )
                                    .padding(.top, 6)
                                    .padding(.bottom, 20)
                                    .padding(.leading, circlePadding + 1)
                                    .animateVmValue(value: state.isRgbSlidersShowed, state: $isRgbSlidersShowed)
                        }
                                .padding(.leading, dividerPadding - circlePadding)
                                .padding(.trailing, sheetHPadding - circlePadding)
                                .safeAreaPadding(.bottom)
                    }
                }

                if (isRgbSlidersShowed) {

                    VStack {

                        MyDivider()

                        ZStack(alignment: .center) {

                            Text(state.rgbText)
                                    .font(.system(size: 16))
                                    .foregroundColor(.white)
                                    .lineLimit(1)
                                    .padding(.horizontal, 8)
                                    .padding(.vertical, 4)
                                    .background(
                                            RoundedRectangle(cornerRadius: 8, style: .continuous)
                                                    .fill(state.selectedColor.toColor())
                                    )

                            HStack {
                                Spacer()
                                Button(
                                        action: { vm.toggleIsRgbSlidersShowed() },
                                        label: {
                                            Image(systemName: "chevron.down.circle.fill")
                                                    .font(.system(size: 24, weight: .medium))
                                                    .foregroundColor(Color(.iconButtonBgBg1))
                                        }
                                )
                            }
                                    .padding(.trailing, sheetHPadding + circlePadding + 2)
                        }
                                .padding(.top, 12)
                                .padding(.bottom, 8)

                        ColorSliderView(value: Double(state.r), color: .red) { vm.upR(r: Float($0)) }
                        ColorSliderView(value: Double(state.g), color: .green) { vm.upG(g: Float($0)) }
                        ColorSliderView(value: Double(state.b), color: .blue) { vm.upB(b: Float($0)) }
                    }
                            .safeAreaPadding(.bottom)
                            .background(Color(.mySecondaryBackground))
                            .transition(.move(edge: .bottom))
                }
            }
        }
                .ignoresSafeArea()
                .background(Color(.mySecondaryBackground))
    }
}

private struct ColorSliderView: View {

    private let valueVM: Double
    @State private var value: Double = 0
    private let color: Color
    private let onChange: (Double) -> Void

    init(
            value: Double,
            color: Color,
            onChange: @escaping (Double) -> Void
    ) {
        _value = State(initialValue: value)
        valueVM = value
        self.color = color
        self.onChange = onChange
    }

    var body: some View {
        Slider(value: $value, in: 0...255)
                ///
                .onChange(of: value) { newValue in
                    onChange(newValue)
                }
                .animateVmValue(value: valueVM, state: $value)
                ///
                .accentColor(color)
                .padding(.horizontal, sheetHPadding)
                .padding(.vertical, 6)
    }
}
